internal/common/tree.go

```diff
@@ -126,19 +182,24 @@ func AddLabelToBranch(node *DNSTreeNode, branch, label, value string) {
 			},
 		})
 	} else {
-		if len(d.Targets) == 1 {
-			d.Labels[label] = value
-		} else {
-			//remove target from shared dataset and recreate uniquely
-			for i, t := range d.Targets {
-				if t == branch {
-					d.Targets = append(d.Targets[:i], d.Targets[i+1:]...)
-					newDS := DNSTreeNodeData{
-						Labels:  d.Labels.DeepCopy(),
-						Targets: []string{branch},
+		for _, ep := range eps {
+			if len(ep.Targets) == 1 {
+				if ep.Labels == nil {
+					ep.Labels = endpoint.NewLabels()
+				}
+				ep.Labels[label] = value
+			} else {
+				//remove target from shared endpoint and recreate uniquely
```

Comment:
What does this mean? Not sure why this would ever be creating an endpoint, it should just be working with endpoints in the tree(referenced from the lists) no?

Reply:
this scenario is very unlikely to occur, but it felt incorrect to leave it unhandled. To recreate it would go something like:
- Have one endpoint: cname.x with targets: cname1.x and cname2.x
- Add a label to cname1.x target

As I see this, we either:
- Add the label to the endpoint that contains the target (thereby inaccurately labelling cname2.x as well)
- Move the target to be labelled to it's own endpoint

In our usecase the label is being used to mark endpoints to be unpublished, so it made sense to me to remove it from the targets and make it in a new endpoint (that will be unpublished) with the label.